Update config_list_from_json function to support YAML parsing

Mai0313 opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Add Yaml file supported for OAI_CONFIG_LIST by simply changing from `json.load` to `yaml.safe_load`

## Related issue number

#2559

## Checks

- [X] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

Mai0313 commented:
> @microsoft-github-policy-service agree

@microsoft-github-policy-service agree company="Mediatek"

Mai0313 commented:
Idk why there is no yaml module, i thought it was a default module for Python

ekzhu reviewed (changes requested):
Thanks for the PR! Could you also update the documentation: website/docs/topics/llm_configuration.ipynb

Mai0313 commented:
> Thanks for the PR! Could you also update the documentation: website/docs/topics/llm_configuration.ipynb

ofc, i will update it once i am at work!

Mai0313 commented:
> website/docs/topics/llm_configuration.ipyn

oh sorry, it is my first time to use `mdx` format, is there anyway I can preview the changes before commit?
I tried [this extension](https://marketplace.visualstudio.com/items?itemName=xyc.vscode-mdx-preview), but it seems not working for me.

Mai0313 commented:
@ekzhu 
It is my first time to use `mdx` format and I do not know where I can preview the changes for `mdx`, so basically, the way I have done is copy-paste the previous content and changed to yaml version.

BTW, I have added `pyyaml` to `setup.py` to make sure no one get error when loading config, since we don't use `json` module but `yaml`.


Additionally, I have implemented pytest for the YAML version. Specifically, I utilized `pytest.mark.parametrize` to minimize code duplication.

ekzhu commented:
> It is my first time to use mdx format and I do not know where I can preview the changes for mdx, so basically, the way I have done is copy-paste the previous content and changed to yaml version.

See https://microsoft.github.io/autogen/docs/contributor-guide/documentation, the `process_notebooks.py` script will generate the MDX files.

ekzhu commented on setup.py:
```diff
@@ -25,6 +25,7 @@
     # Disallowing 2.6.0 can be removed when this is fixed https://github.com/pydantic/pydantic/issues/8705
     "pydantic>=1.10,<3,!=2.6.0",  # could be both V1 and V2
     "docker",
+    "pyyaml",
```
We can add `pyyaml` to test dependencies. But for install dependency let's move this to an optional dependency like those below.

Mai0313 replied:
actually I dont see any of test dependency or optional dependency under setup.py
should I put it under test from extra_require?

thinkall replied:
You can add it like here: https://github.com/microsoft/autogen/blob/82903f5f89a4f514b267911e56c3032b339f2581/setup.py#L60-L70

ekzhu commented on autogen/oai/openai_utils.py:
```diff
@@ -504,11 +522,11 @@ def config_list_from_json(
             # Else, it should be a JSON string by itself.
-            json_str = env_str
-        config_list = json.loads(json_str)
+            json_or_yaml_str = env_str
```
If the yaml dependency is not installed, we can fall back to use `json` module. We can do this by checking for yaml import error during import, and use that information to choose whether to use yaml load or json load.

Mai0313 replied:
oh, it sounds good

ekzhu commented on autogen/oai/openai_utils.py:
```diff
@@ -475,21 +476,38 @@ def config_list_from_json(
     Args:
         env_or_file (str): The name of the environment variable, the filename, or the environment variable of the filename
-            that containing the JSON data.
+            that containing the JSON/YAML data.
```
Add explanation for installing `pyyaml`.

thinkall commented on autogen/oai/openai_utils.py:
```diff
@@ -503,11 +530,11 @@ def config_list_from_json(
         if os.path.exists(env_str):
             # It is a file location, and we need to load the json from the file.
```
```suggestion
            # It is a file location, and we need to load the json/yaml from the file.
```

thinkall commented on autogen/oai/openai_utils.py:
```diff
@@ -503,11 +530,11 @@ def config_list_from_json(
-                json_str = file.read()
+                json_or_yaml_str = file.read()
         else:
             # Else, it should be a JSON string by itself.
```
```suggestion
            # Else, it should be a JSON/YAML string by itself.
```

thinkall reviewed (commented):
Thank you very much for the PR, @Mai0313 . I've left a few comments, but it overall looks good to me.

thinkall commented:
Some tests failed,  could you fix them? For documentation, you can follow https://microsoft.github.io/autogen/docs/contributor-guide/documentation to check it locally.
